django/cantusdb_project/main_app/views/views.py

```diff
@@ -620,3 +626,64 @@ def content_overview(request):
     return render(
         request, "content_overview.html", {"objects": recently_updated_50_objects}
     )
+
+def redirect_node_url(request, pk: int) -> HttpResponse:
+    """
+    A function that will redirect /node/ URLs from OldCantus to their corresponding page in NewCantus.
+    This makes NewCantus links backwards compatible for users who may have bookmarked these types of URLs in OldCantus.
+    In addition, this function (paired with get_user_id() below) account for the different numbering systems in both versions of CantusDB, notably for /indexer/ paths which are now at /user/.
+
+    Takes in a request and the primary key (ID following /node/ in the URL) as arguments.
+    Returns the matching page in NewCantus if it exists and a 404 otherwise.
+    """
+    not_found = HttpResponseNotFound()
+    
+    # all IDs above this value are created in NewCantus and thus could have conflicts between types.
+    # this number is a placeholder and will be updated post-migration.
+    # we will manually create (unpublished) dummy objects in the database to ensure that all subqequent objects created will have IDs above this number.
+    if pk >= 1_000_000:
+        return not_found
+    
+    # chant, source, sequence, article
+    possible_types = [
+        (Chant, 'chant-detail'),
+        (Source, 'source-detail'),
+        (Sequence, 'sequence-detail'),
+        (Article, 'article-detail')
+    ]
+
+    user_id = get_user_id(pk)
+    if get_user_id(pk) is not None:
+        return redirect('user-detail', user_id)
+    
+    for (rec_type, view) in possible_types:
+        if record_exists(rec_type, pk):
+            # if an object is found, a redirect() call to the appropriate view is returned
+            return redirect(view, pk)
+        
+    # if it reaches the end of the types with finding an existing object, a 404 will be returned
+    return not_found
+
+# used to determine whether record of specific type (chant, source, sequence, article) exists for a given pk
+def record_exists(rec_type: BaseModel, pk: int) -> bool:
+    try:
+        result = rec_type.objects.get(id=pk)
+        print(f'result: {result}')
+        return True
+    except rec_type.DoesNotExist:
+        return False
+
+def get_user_id(pk: int) -> Optional[int]:
```

Comment:
`def get_user_id_from_old_indexer_id(indexer_id: int)` or even `def get_user_id_from_indexer_id(...)` would make it clearer what this function does.

Reply:
changed